ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -114,7 +114,86 @@ public void start() throws StoreException {
+    try {
+      List<BlobId> blobIdList = ids.stream().map(key -> (BlobId) key).collect(Collectors.toList());
+      Map<String, CloudBlobMetadata> cloudBlobMetadataListMap = cloudDestination.getBlobMetadata(blobIdList);
+      if (cloudBlobMetadataListMap.size() < blobIdList.size()) {
+        Set<BlobId> missingBlobs = blobIdList.stream()
+            .filter(blobId -> !cloudBlobMetadataListMap.containsKey(blobId))
+            .collect(Collectors.toSet());
+        throw new StoreException("Some of the keys were missing in the cloud metadata store: " + missingBlobs,
+            StoreErrorCodes.ID_Not_Found);
+      }
+      validateCloudMetadata(cloudBlobMetadataListMap, storeGetOptions);
+      for (BlobId blobId : blobIdList) {
+        CloudBlobMetadata blobMetadata = cloudBlobMetadataListMap.get(blobId.getID());
+        MessageInfo messageInfo = new MessageInfo(blobId, blobMetadata.getSize(), blobMetadata.getExpirationTime(),
+            (short) blobMetadata.getAccountId(), (short) blobMetadata.getContainerId(), getOperationTime(blobMetadata));
+        messageInfos.add(messageInfo);
+        blobReadInfos.add(new CloudMessageReadSet.BlobReadInfo(blobMetadata, blobId));
+      }
+    } catch (CloudStorageException e) {
+      throw new StoreException(e, StoreErrorCodes.IOError);
+    }
+    CloudMessageReadSet messageReadSet = new CloudMessageReadSet(blobReadInfos, cloudDestination);
+    return new StoreInfo(messageReadSet, messageInfos);
+  }
+
+  /**
+   * validate the {@code CloudBlobMetadata} map to make sure it has metadata for all keys, and they meet the {@code storeGetOptions} requirements.
+   * @param cloudBlobMetadataListMap
+   * @throws StoreException if the {@code CloudBlobMetadata} isnt valid
+   */
+  private void validateCloudMetadata(Map<String, CloudBlobMetadata> cloudBlobMetadataListMap,
+      EnumSet<StoreGetOptions> storeGetOptions) throws StoreException {
+    for (String key : cloudBlobMetadataListMap.keySet()) {
+      if (isBlobDeleted(cloudBlobMetadataListMap.get(key)) && !storeGetOptions.contains(
+          StoreGetOptions.Store_Include_Deleted)) {
+        throw new StoreException("Id " + key + " has been deleted on the cloud", StoreErrorCodes.ID_Deleted);
+      }
+      long currentTimestamp = System.currentTimeMillis();
+      if (isBlobExpired(cloudBlobMetadataListMap.get(key), currentTimestamp) && !storeGetOptions.contains(
+          StoreGetOptions.Store_Include_Expired)) {
+        throw new StoreException("Id " + key + " has expired on the cloud", StoreErrorCodes.TTL_Expired);
+      }
+    }
+  }
+
+  /**
+   * Gets the operation time for a blob from blob metadata based on the blob's current state and timestamp recorded for that state.
+   * @param metadata blob metadata from which to derive operation time.
+   * @return operation time.
+   */
+  private long getOperationTime(CloudBlobMetadata metadata) {
+    long currentTimeStamp = System.currentTimeMillis();
```

Comment:
We have the same issue in this method with System.currentTimeMillis() being called in a loop.  Please call it once at the beginning of the get method and pass to all utility methods that need it.

Reply:
fixed.